Tasks/NuGetInstaller/VsoNuGetHelper.ps1

```diff
@@ -0,0 +1,157 @@
+Add-Type -AssemblyName System.Security
+
+$nuGetTempDirectory = Join-Path $Env:AGENT_BUILDDIRECTORY "NuGet\"
+$tempNuGetConfigPath = Join-Path $nuGetTempDirectory "newNuGet.config"
+
+function SaveTempNuGetConfig
+{
+    param(
+        [xml] $nugetConfig        
+    )
+
+    if(-not (Test-Path -Path ($nuGetTempDirectory)))
+    {
+        New-Item -Path $nuGetTempDirectory -ItemType Directory
+    }
+    Write-Output "Saving to $tempNuGetConfigPath"
+    $nugetConfig.Save($tempNuGetConfigPath)
+}
+
+
+function EncryptNuGetPassword
+{
+    param(
+        [string] $password
+    )
+
+    [Convert]::ToBase64String(
+        [System.Security.Cryptography.ProtectedData]::Protect(
+            [Text.Encoding]::UTF8.GetBytes($password),
+            [Text.Encoding]::UTF8.GetBytes("NuGet"),
+            [System.Security.Cryptography.DataProtectionScope]::CurrentUser));
+}
+
+function SetNewCredentialElement
+{
+    param(
+        [xml] $nugetConfig,
+        [System.Xml.XmlElement] $credentialsSection,
+        [string] $credentialKey,
+        [string] $password
+    )
+
+    $newCredential = $nugetConfig.CreateElement($credentialKey)
+
+    $userNameElement= $nugetConfig.CreateElement("add")
+    [void]$userNameElement.SetAttribute("key", "UserName")
+    [void]$userNameElement.SetAttribute("value", "VssSessionToken")
+    [void]$newCredential.AppendChild($userNameElement)
+
+    $passwordElement = $nugetConfig.CreateElement("add")
+    [void]$passwordElement.SetAttribute("key", "Password")
+    $encryptedPassword = EncryptNuGetPassword $password
+    [void]$passwordElement.SetAttribute("value", $encryptedPassword)
+    [void]$newCredential.AppendChild($passwordElement)
+
+    [void]$credentialsSection.AppendChild($newCredential)
+}
+
+function ShouldGetCredentialsForFeed
+{
+    param(
+        [Uri]$feedUri
+    )
+
+    if(-not ($feedUri.Scheme.Equals("http","InvariantCultureIgnoreCase") -or  $feedUri.Scheme.Equals("https","InvariantCultureIgnoreCase")))
+    {
+        return $false
+    }
+
+    $WebRequestObject = [System.Net.HttpWebRequest] [System.Net.WebRequest]::Create($feedUri);
+    try
+    {
+        $ResponseObject = [System.Net.HttpWebResponse] $WebRequestObject.GetResponse();
+    }
+    catch [Net.WebException] 
+    {
+        $ResponseObject = [System.Net.HttpWebResponse] $_.Exception.Response
+    }
+    
+    if($ResponseObject.StatusCode -eq 401)
+    {
+        $tfsHeaderPresent = $false
+        $vssHeaderPresent = $false
+        foreach($key in $ResponseObject.Headers.AllKeys)
+        {
+            if($tfsHeaderPresent -or $key.Contains("X-TFS"))
```

Comment:
It is a minor optimization to avoid a string match when it has already been found.